lib/wibox/container/border.lua

```diff
@@ -0,0 +1,1050 @@
+-- image slicing or widgets within the border. The `awful.layout.align` layout
+-- an also be used to achieve border, but make it very hard to make them
+-- symmetric on both axis.
+--
+-- Note that because of legacy reasons, `wibox.container.background` also has
+-- good support for borders. If you only need simple shaped strokes, the,
+-- the `background` container is a much better choice. This module is better
+-- suited for background images and border widgets.
+--
+-- Advanced usage
+-- ==============
+--
+-- In this first example, 2 `wibox.container.borders` are combined in a
+-- `wibox.layout.stack`. The bottom one is the "real" border and the top
+-- layer is used to place some widgets. This is useful to built titlebars.
+--
+-- @DOC_wibox_container_border_dual1_EXAMPLE@
+--
+-- This example demonstrates how to use this module to create a client
+-- border with a top titlebar and borders. It does so by embedding a
+-- `wibox.container.border` into another `wibox.container.border. The outer
+-- container acts as the border around the central area while the inner one
+-- spans the top area and contains the titlebar itself. The outer border
+-- widgets area can be used to implement border resize.
+--
+-- @DOC_wibox_container_border_titlebar1_EXAMPLE@
+--
+-- This container is like every other widgets. It is possible to override the
+-- Cairo drawing functions. This is very useful here to create custom borders
+-- using cairo while still re-using the internal geometry handling code.
+--
+-- @DOC_wibox_container_border_custom_draw1_EXAMPLE@
+--
+--@DOC_wibox_container_defaults_border_EXAMPLE@
+--
+-- @author Emmanuel Lepage-Vallee
+-- @copyright 2021 Emmanuel Lepage-Vallee
+-- @containermod wibox.container.border
+-- @supermodule wibox.widget.base
+local gtable   = require("gears.table")
+local imagebox = require("wibox.widget.imagebox")
+local base     = require("wibox.widget.base")
+local gsurface = require("gears.surface")
+local cairo    = require("lgi").cairo
+
+local components = {
+    "top_left", "top", "top_right", "right", "bottom_right", "bottom",
+    "bottom_left", "left", "fill"
+}
+
+local slice_modes = {
+    top_left     = "corners",
+    top          = "sides",
+    top_right    = "corners",
+    right        = "sides",
+    bottom_right = "corners",
+    bottom       = "sides",
+    bottom_left  = "corners",
+    left         = "sides",
+    fill         = "filling"
+}
+
+local fit_types = { "corners", "sides", "filling" }
+
+local module = {}
+
+local function imagebox_fit(self, _, w, h)
+    return math.min(w, self.source_width or 0), math.min(h, self.source_height or 0)
+end
+
+local function fit_common(widget, ctx, max_w, max_h)
+    if not widget then return 0, 0 end
+
+    local w, h = widget:fit(ctx, max_w or math.huge, max_h or math.huge)
+
+    return w, h
+end
+
+local function uses_slice(self)
+    return (self._private.border_widgets or self._private.border_image_widgets) == nil
```

Comment:
```suggestion
    return not (self._private.border_widgets or self._private.border_image_widgets)
```

Sounds easier to read. Or am I missing something?